src/node_api_embedding.cc

```diff
@@ -0,0 +1,254 @@
+    if (err_handler != nullptr) {
+      err_handler(error.c_str());
+    } else {
+      fprintf(stderr, "%s\n", error.c_str());
+    }
+  }
+  if (setup == nullptr) {
+    return napi_generic_failure;
+  }
+
+  std::shared_ptr<node::StaticExternalTwoByteResource> main_resource = nullptr;
+  if (main_script != nullptr) {
+    // We convert the user-supplied main_script to a UTF-16 resource
+    // and we store its shared_ptr in the environment
+    size_t u8_length = strlen(main_script);
+    size_t expected_u16_length =
+        simdutf::utf16_length_from_utf8(main_script, u8_length);
+    auto out = std::make_shared<std::vector<uint16_t>>(expected_u16_length);
+    size_t u16_length = simdutf::convert_utf8_to_utf16(
+        main_script, u8_length, reinterpret_cast<char16_t*>(out->data()));
+    out->resize(u16_length);
+    main_resource = std::make_shared<node::StaticExternalTwoByteResource>(
+        out->data(), out->size(), out);
+  }
+
+  auto emb_env =
+      new v8impl::EmbeddedEnvironment(std::move(setup), main_resource);
+
+  std::string filename =
+      wrapper->args().size() > 1 ? wrapper->args()[1] : "<internal>";
+  auto env__ =
+      new node_napi_env__(emb_env->setup()->context(), filename, api_version);
+  emb_env->setup()->env()->set_embedded(emb_env);
+  env__->node_env()->AddCleanupHook(
+      [](void* arg) { static_cast<napi_env>(arg)->Unref(); },
+      static_cast<void*>(env__));
+  *result = env__;
+
+  auto env = emb_env->setup()->env();
+
+  auto ret = node::LoadEnvironment(
+      env,
+      [env, resource = main_resource.get()](
+          const node::StartExecutionCallbackInfo& info)
+          -> v8::MaybeLocal<v8::Value> {
+        node::Realm* realm = env->principal_realm();
+        auto ret = realm->ExecuteBootstrapper(
+            "internal/bootstrap/switches/is_embedded_env");
+        if (ret.IsEmpty()) return ret;
+
+        std::string name =
+            "embedder_main_napi_" + std::to_string(env->thread_id());
+        if (resource != nullptr) {
+          env->builtin_loader()->Add(name.c_str(), node::UnionBytes(resource));
+          return realm->ExecuteBootstrapper(name.c_str());
+        } else {
+          return v8::Undefined(env->isolate());
+        }
+      });
+  if (ret.IsEmpty()) return napi_pending_exception;
+
+  emb_env->seal();
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL napi_destroy_environment(napi_env env, int* exit_code) {
+  CHECK_ARG(env, env);
+  node_napi_env node_env = reinterpret_cast<node_napi_env>(env);
+
+  int r = node::SpinEventLoop(node_env->node_env()).FromMaybe(1);
+  if (exit_code != nullptr) *exit_code = r;
+  node::Stop(node_env->node_env());
+
+  auto emb_env = reinterpret_cast<v8impl::EmbeddedEnvironment*>(
+      node_env->node_env()->get_embedded());
+  node_env->node_env()->set_embedded(nullptr);
+  // This deletes the uniq_ptr to node::CommonEnvironmentSetup
+  // and the v8::locker
+  delete emb_env;
+
+  cppgc::ShutdownProcess();
```

Comment:
This call to `ShutdownProcess()` prevents another environment instance from being created in the same process. Specifically, a second call to `napi_create_environment()` crashes the process:
```
# Check failed: (g_page_allocator) != nullptr.
``` 
I found this because the [node-api-dotnet project](https://github.com/microsoft/node-api-dotnet/pull/219) runs multiple embedding test cases in a process, and each test case uses a separate environment instance. With this line removed, the latest PR revision works well.

Was there a reason to add this call to `ShutdownProcess()`?